File: tests/validate-refs.py
#!/usr/bin/env python3
"""Static reference validator for Dora Fares Al Shamal frontend.

Checks every *.html file for broken local references (scripts, styles,
images, links) and inline <script> blocks. Does NOT require a server.

Usage:
    python tests/validate-refs.py
"""
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
HTML = sorted(ROOT.glob("*.html"))
IGNORED_LINK_PREFIXES = (
    "http://", "https://", "//", "data:", "#", "mailto:", "tel:",
    "javascript:",
)


def strip_fragment_query(url: str) -> str:
    return url.split("?")[0].split("#")[0]


def collect_refs(html_text: str, attr: str):
    pattern = re.compile(rf"{attr}\s*=\s*['\"]([^'\"]+)['\"]", re.IGNORECASE)
    return pattern.findall(html_text)


def main():
    errors = []
    inline_script_pages = []

    for page in HTML:
        text = page.read_text(encoding="utf-8", errors="ignore")

        # Detect inline <script> blocks (non src)
        if re.search(
            r'<script\s*(?![^>]*src=)(?![^>]*type\s*=\s*["\']application/ld\+json["\'])[^>]*>',
            text,
            re.IGNORECASE,
        ):
            inline_script_pages.append(page.name)

        # Scripts
        for src in collect_refs(text, "src"):
            if src.startswith(IGNORED_LINK_PREFIXES):
                continue
            clean = strip_fragment_query(src)
            if clean and not (ROOT / clean).exists():
                errors.append(f"{page.name}: missing script '{src}'")

        # Stylesheets / links
        for href in collect_refs(text, "href"):
            if href.startswith(IGNORED_LINK_PREFIXES):
                continue
            clean = strip_fragment_query(href)
            if not clean:
                continue
            target = ROOT / clean
            if not target.exists():
                # Anchor-only links to existing pages are OK
                if clean.endswith(".html") and not target.exists():
                    errors.append(f"{page.name}: missing link '{href}'")
                elif not clean.endswith(".html") and not target.exists():
                    errors.append(f"{page.name}: missing asset '{href}'")

    print(f"Checked {len(HTML)} HTML pages.")

    if inline_script_pages:
        print("\n⚠️  Pages with inline <script> blocks:")
        for p in inline_script_pages:
            print(f"  - {p}")
    else:
        print("\n✅ No inline <script> blocks found.")

    if errors:
        print(f"\n❌ {len(errors)} broken reference(s) found:")
        for e in errors:
            print(f"  - {e}")
        sys.exit(1)
    else:
        print("\n✅ No broken local references found.")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: tests/validate-refs.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;
use Cwd 'abs_path';

my $root = dirname(dirname(abs_path($0)));
chdir $root or die "Cannot chdir $root: $!";

my @html = sort glob("*.html");
my @errors;
my @inline;

my @ignored = ('http://', 'https://', '//', 'data:', '#', 'mailto:', 'tel:', 'javascript:');

sub strip_qf {
    my $u = shift;
    $u =~ s/\?.*//;
    $u =~ s/#.*//;
    return $u;
}

sub is_external {
    my $u = shift;
    for my $p (@ignored) {
        return 1 if index($u, $p) == 0;
    }
    return 0;
}

for my $page (@html) {
    open my $fh, '<:encoding(UTF-8)', $page or die $!;
    local $/;
    my $text = <$fh>;
    close $fh;

    if ($text =~ /<script\s+(?![^>]*src=)(?![^>]*type\s*=\s*["']application\/ld\+json["'])[^>]*>/is) {
        push @inline, $page;
    }

    while ($text =~ /src\s*=\s*["']([^"']+)["']/gis) {
        my $src = $1;
        next if is_external($src);
        my $clean = strip_qf($src);
        next unless $clean;
        push @errors, "$page: missing script '$src'" unless -f $clean;
    }

    while ($text =~ /href\s*=\s*["']([^"']+)["']/gis) {
        my $href = $1;
        next if is_external($href);
        my $clean = strip_qf($href);
        next unless $clean;
        if ($clean =~ /\.html$/) {
            push @errors, "$page: missing link '$href'" unless -f $clean;
        } else {
            push @errors, "$page: missing asset '$href'" unless -e $clean;
        }
    }
}

print "Checked ", scalar(@html), " HTML pages.\n";

if (@inline) {
    print "\n⚠️  Pages with inline <script> blocks:\n";
    print "  - $_\n" for @inline;
} else {
    print "\n✅ No inline <script> blocks found.\n";
}

if (@errors) {
    print "\n❌ ", scalar(@errors), " broken reference(s) found:\n";
    print "  - $_\n" for @errors;
    exit 1;
} else {
    print "\n✅ No broken local references found.\n";
    exit 0;
}
